lib/Conversion/TorchOnnxToTorch/OnnxLstmExpander.cpp

```diff
@@ -0,0 +1,479 @@
+  Value cstBatchSize = rewriter.create<Torch::ConstantIntOp>(
+      loc, intType,
+      rewriter.getIntegerAttr(rewriter.getIntegerType(64), batch_size));
+  Value cstHiddenSize = rewriter.create<Torch::ConstantIntOp>(
+      loc, intType,
+      rewriter.getIntegerAttr(rewriter.getIntegerType(64), hidden_size));
+  Value cstNone = rewriter.create<Torch::ConstantNoneOp>(loc);
+  Value cstZero = rewriter.create<Torch::ConstantIntOp>(
+      loc, intType, rewriter.getIntegerAttr(rewriter.getIntegerType(64), 0));
+  Value cstOne = rewriter.create<Torch::ConstantIntOp>(
+      loc, intType, rewriter.getIntegerAttr(rewriter.getIntegerType(64), 1));
+
+  Value HShape = rewriter.create<Torch::PrimListConstructOp>(
+      loc, rewriter.getType<Torch::ListType>(intType),
+      ValueRange({cstNumDirections, cstBatchSize, cstHiddenSize}));
+
+  Value cstDtype =
+      Torch::getDtypeIntValueForType(rewriter, loc, XType.getDtype());
+
+  // initialize hidden and cell states
+  if (binder.tensorOperandAtIndex(initial_h, 5)) {
+    initial_h = rewriter.create<Torch::AtenZerosOp>(
+        loc, HType, HShape, cstDtype, cstNone, cstNone, cstNone);
+  }
+  if (binder.tensorOperandAtIndex(initial_c, 6)) {
+    initial_c = rewriter.create<Torch::AtenZerosOp>(
+        loc, HType, HShape, cstDtype, cstNone, cstNone, cstNone);
+  }
+
+  Value initial_h_forward = getDirection(0, initial_h);
+  Value initial_c_forward = getDirection(0, initial_c);
+
+  // ### everything hereon is only one direction. they won't have the direction
+  // dimension. todo: support bidirectional and reverse LSTM,
+  // possibly by doing each direction separately and combining
+  Value HSizeX4 = rewriter.create<Torch::ConstantIntOp>(
+      loc, intType,
+      rewriter.getIntegerAttr(rewriter.getIntegerType(64), 4 * hidden_size));
+  Value HSizeX8 = rewriter.create<Torch::ConstantIntOp>(
+      loc, intType,
+      rewriter.getIntegerAttr(rewriter.getIntegerType(64), 8 * hidden_size));
+
+  Torch::ValueTensorType biasType = rewriter.getType<Torch::ValueTensorType>(
+      llvm::SmallVector<int64_t>{hidden_size * 4}, WType.getDtype());
+
+  Value Wb = rewriter.create<Torch::AtenSliceTensorOp>(
+      loc, biasType,
+      /*input=*/B_forward, /*dim=*/cstZero, /*start=*/cstZero,
+      /*end=*/HSizeX4, /*step=*/cstOne);
+  Value Rb = rewriter.create<Torch::AtenSliceTensorOp>(
+      loc, biasType,
+      /*input=*/B_forward, /*dim=*/cstZero, /*start=*/HSizeX4,
+      /*end=*/HSizeX8, /*step=*/cstOne);
+
+  auto [Y_nonumdirections, Y_h, Y_c] =
+      lstm_layer(binder, rewriter, X, initial_h_forward, initial_c_forward,
+                 W_forward, Wb, R_forward, Rb, activations);
+
+  // ### everything hereon has to have the direction dimension again ###
+  // unsqueeze dim0 of Y_H and Y_c
+  // Y_h = Y_h.unsqueeze(0)
+  // Y_c = Y_c.unsqueeze(0)
+  assert(num_directions == 1); // TODO: support bidirectional LSTM by doing both
```

Comment:
You should not be doing an assertion for something unsupported, you should be returning a match failure. You should also be checking this before you do the construction.